### PR TITLE
UCS/BITMAP/TEST: Implement find-first, find-next and foreach

### PR DESCRIPTION
**Pull request by yosefe**
## Why
Extend the implementation of dynamic bitmap to support bit iteration

### CODE REVIEW

File: src/ucs/datastruct/bitmap.h [hunk 1]
```diff
@@ -597,6 +597,44 @@ _UCS_BITMAP_DECLARE_TYPE(128)
 _UCS_BITMAP_DECLARE_TYPE(256)
 
 
+/**
+ * Helper macro to iterate over all set (1) bits of a given bitmap.
+ *
+ * @param _bit_index Bit index (global offset - relative to the whole bitmap)
+ * @param _bits      Iterate over bits of this bitmap
```
Comment (gleon99) [inline src/ucs/datastruct/bitmap.h:604]: \+ `_num_words`

File: src/ucs/datastruct/bitmap.h [hunk 2]
```diff
@@ -597,6 +597,44 @@ _UCS_BITMAP_DECLARE_TYPE(128)
...
+         _bit_index < (_num_words)*UCS_BITMAP_BITS_IN_WORD; \
+         _bit_index = ucs_bitmap_bits_ffs(_bits, _num_words, _bit_index + 1))
+
+
+/**
+ * Helper function to find the index of the first bit set to 1 in a given
+ * bitmap, starting from a particular index (inclusive). If all bits are zero,
+ * returns the index past the last bit (bitmap size).
+ *
+ * @param bitmap_words Look for the first bit in the words of this bitmap
```
Comment (gleon99) [inline src/ucs/datastruct/bitmap.h:617]: -> `bits`.

File: .clang-format [hunk 3]
```diff
@@ -68,6 +68,7 @@ SpaceBeforeAssignmentOperators: true
 SpaceAfterCStyleCast: false
 SortIncludes: false
 ForEachMacros: ['_UCS_BITMAP_FOR_EACH_WORD',
+                'UCS_DYNAMIC_BITMAP_FOR_EACH_BIT',
```
Comment (gleon99) [inline .clang-format:71]: + `UCS_BITMAP_BITS_FOR_EACH_BIT` ?

File: src/ucs/datastruct/bitmap.c [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    size_t word_index = start_index / UCS_BITMAP_BITS_IN_WORD;
+    size_t mask       = ~(ucs_bitmap_word_bit_mask(start_index) - 1);
+    ucs_bitmap_word_t masked_word;
+    unsigned bit_index;
+
+    UCS_BITMAP_CHECK_INDEX(start_index, num_words, <=);
+    while (word_index < num_words) {
+        masked_word = bits[word_index] & mask;
+        while (masked_word != 0) {
+            /* TODO optimize by skipping multiple bits */
```
Comment (gleon99) [inline src/ucs/datastruct/bitmap.c:23]: Maybe impl it before merge?
Having nested loop for a single pass over all bits is a bit weird..

File: src/ucs/datastruct/bitmap.h [hunk 5]
```diff
@@ -597,6 +597,44 @@ _UCS_BITMAP_DECLARE_TYPE(128)
...
+
+
+/*
+ * Helper function to find the index of the n-th bit set to 1 in a given bitmap,
+ * starting from a particular index (inclusive). If all bits are zero, returns
+ * the index past the last bit (bitmap size).
+ *
+ * @param bitmap_words Look for the first bit in the words of this bitmap
+ * @param num_words    Number of words in the bitmap
+ * @param start_index  The first bit to look from
```
Comment (gleon99) [inline src/ucs/datastruct/bitmap.h:632]: \+ `n`.
Also, maybe rename it to smth more descriptive?

File: test/gtest/ucs/test_bitmap.cc [hunk 6]
```diff
@@ -412,6 +412,43 @@ UCS_TEST_F(test_dynamic_bitmap, test_popcount_upto_index)
     EXPECT_EQ(1, popcount);
 }
 
+UCS_TEST_F(test_dynamic_bitmap, test_ffs_fns) {
```
Comment (iyastreb) [inline test/gtest/ucs/test_bitmap.cc:415]: maybe consider adding tests for corner cases:
- bitmap of size 100, all bits are unset
- bitmap of size 100, all bits are set
- bitmap of size 0/1

**Review by iyastreb — DISMISSED** [review]

File: src/ucs/datastruct/bitmap.c [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+                           size_t start_index)
+{
+    return ucs_bitmap_bits_fns(bits, num_words, start_index, 0);
+}
+
+size_t ucs_bitmap_bits_fns(const ucs_bitmap_word_t *bits, size_t num_words,
+                           size_t start_index, size_t bit_count)
+{
+    return ucs_bitmap_bits_fns_inline(bits, num_words, start_index, bit_count);
+}
```
Comment (brminich) [inline src/ucs/datastruct/bitmap.c:72]: do we need them to return `size_t`, not `int`?

Comment (brminich) [reply src/ucs/datastruct/bitmap.c:76]: unresolved,
pls note that we return `int` from `ucs_dynamic_bitmap_ffs` and `ucs_dynamic_bitmap_fns`

File: src/ucs/datastruct/bitmap.h [hunk 8]
```diff
@@ -597,6 +597,46 @@ _UCS_BITMAP_DECLARE_TYPE(128)
...
+
+/**
+ * Helper function to find the index of the @a bit_count-th bit set to 1 in a
+ * given bitmap, starting from the index @a start_index (inclusive). If all bits
+ * are zero, returns the index past the last bit (bitmap size).
+ *
+ * @param bits         Look for the first bit in the words of this bitmap
+ * @param num_words    Number of words in the bitmap
+ * @param start_index  The first bit to look from
+ * @param bit_count    Number of bits to look for.
```
[the remark this answers is not in the document]
Comment (brminich) [reply src/ucs/datastruct/bitmap.h:634]: unresolved

File: src/ucs/datastruct/bitmap.c (hunk 7 above)
Comment (yosefe) [reply src/ucs/datastruct/bitmap.c:76]: converted all to size_t

**Review by brminich — APPROVED** [review]

File: src/ucs/datastruct/bitmap.h [hunk 9]
```diff
@@ -597,6 +597,46 @@ _UCS_BITMAP_DECLARE_TYPE(128)
...
+
+/**
+ * Helper function to find the index of the @a bit_count-th bit set to 1 in a
+ * given bitmap, starting from the index @a start_index (inclusive). If all bits
+ * are zero, returns the index past the last bit (bitmap size).
+ *
+ * @param bits         Look for the first bit in the words of this bitmap.
+ * @param num_words    Number of words in the bitmap.
+ * @param start_index  The first bit to look from.
+ * @param bit_count    Number of bits to look for.
```
Comment (tvegas1) [inline src/ucs/datastruct/bitmap.h:634]: Seems this one can be called with `bit_count=0`, shall we improve this documentation by adding acceptable range values? Seems confusing to have `Number of bits to look for` being zero. Or maybe rename to `last_bit`/`bit_skip`?

Comment (yosefe) [reply src/ucs/datastruct/bitmap.h]: @tvegas1 fixed

**Review by brminich — APPROVED** [review]
